### ISSUE TITLE
Assertion in `fgArgInfo::RemorphRegArg` when running CoreFX tests with a debug CoreCLR

### ISSUE THREAD

**Issue by pgavlin** [dotnet/coreclr#1995]
```
Assert failure(PID 25763 [0x000064a3], Thread: 25763 [0x64a3]): Assertion failed 'curArgTabEntry->regNum == regNum' in 'System.Linq.Parallel.AssociativeAggregationOperator`3[Nullable`1,Pair`2,Nullable`1][System.Nullable`1[System.Single],System.Linq.Parallel.Pair`2[System.Boolean,System.Nullable`1[System.Single]],System.Nullable`1[System.Single]]:Aggregate():struct:this' (IL size 185)

    File: /home/pgavlin/dev/coreclr/src/jit/morph.cpp Line: 1285
    Image: /home/pgavlin/dev/corefx/bin/tests/Linux.AnyCPU.Debug/System.Linq.Parallel.Tests/dnxcore50/corerun
```

I believe that the assertion is due to the following situation:
- A struct S with SysV classification (INTEGER, SSE) is passed by value to a call before another argument with SysV classification INTEGER
    - `fgArgInfo.numRegs` is 2 for the argument, since it takes one integer and one SSE register
- During `fgMorph`, the struct argument is morphed from an ldobj into a copyblk to a temporary for argument setup (it's not clear why this is necessary; more on this later), and a list of (lclFld longField, lclFld floatField) that replaces the original argument
- During assertion propagation (or other optimizations), the call arguments are remorphed
- During remorphing, the first struct argument (or perhaps only the first `lclFld`) ends up accounting for two integer registers, since its size is calculated using `fgArgInfo.numRegs` for the argument info for the original ldobj. This in turn increments the integer register number by 2, causing the mismatch mentioned in the assert when processing the next argument.

Regarding the potentially unnecessary copyblk: the copyblk is caused by the [following code in fgMorphArgs](https://github.com/dotnet/coreclr/blob/master/src/jit/morph.cpp#L3116):

```
                        if (!structDesc.passedInRegisters)
                        {
                            passStructByRef = false;
                            copyBlkClass = NULL;
                        }
                        else 
                        {
                            passStructByRef = true;
                            copyBlkClass = ldObjClass;
                        }
```

The condition on the `if` statement seems backwards to me, since it seems to imply that structs that have been classified as passed in registers should actually be passed by implicit reference. Unfortunately, a naive inversion of this condition results in further asserts in other SysV-related code.

**Comment by pgavlin** [comment]
Compiling and running the following program on a debug JIT will demonstrate the issue:

```
using System;
using System.Runtime.CompilerServices;

struct S
{
	public int i;
	public int j;
	public float f;
}

sealed class C
{
	public S s;
}

static class Repro
{
	[MethodImpl(MethodImplOptions.NoInlining)]
	static S Accept(S a, S b)
	{
		return b;
	}

	static float M(C c)
	{
		var s = default(S);
		if (c != null)
		{
			s = Accept(c.s, c.s);
		}

		return s.f;
	}

	static int Main()
	{
		return (int)M(new C());
	}
}
```

**Comment by pgavlin** [comment]
AFAICT, this does not cause SBCG on release builds; just an assert on debug builds.

**Comment by pgavlin** [comment]
@LLITCHEV FYI.

**Comment by LLITCHEV** [comment]
Does this happen without the changes in PR 1981? The if condition above looks fine. I ran few corefx and coreclr tests over the weekend and they all passed with no problems.

**Comment by pgavlin** [comment]
Yes, this happens without the changes in #1981.

**Comment by pgavlin** [comment]
> The if condition above looks fine.

It very well may be, but can you please explain why? If it is correct, it at least deserves a comment, as it appears quite counter-intuitive.

**Comment by pgavlin** [comment]
Closed by #2009.
